File: crates/spoke-connect/src/hello.rs
//! Authenticated hello: transport adapter for `spoke-connect-hello-jcs-v1`.
//!
//! The signing/verification rules themselves live in the pure session core
//! ([`crate::core::sign_hello_ed25519`] / [`crate::core::verify_hello_ed25519`]
//! over raw Ed25519 key bytes). This module is the **boundary adapter**: it
//! converts libp2p identity types to raw key material and maps core errors to
//! transport errors.
//!
//! Normative rules (`.mstar/specs/spoke-connect.md`):
//! - Signed object = exactly `{protocol_version, peer_id, nonce, host}`.
//! - Canonicalize with RFC 8785 JCS (via `serde_jcs`), sign with the libp2p
//!   identity keypair (Ed25519), encode raw signature bytes base64url (no padding).
//! - Verification uses the public key bound to the **noise-authenticated**
//!   remote peer (libp2p identify carries it); the claimed `peer_id` in the
//!   hello must equal that authenticated peer id.
//! - Top-level hello `extensions` and the `signature` field are not part of
//!   the signed object.

use crate::core;
use crate::error::{map_core_error, ConnectError};
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use libp2p::identity::{Keypair, PublicKey};
use libp2p::PeerId;
use spoke_schemas::connect::connect_hello::HostCapabilityManifest;
use spoke_schemas::connect::ConnectHello;

/// Generate a single-use hello nonce: 128 bits of CSPRNG entropy, base64url
/// encoded (22 characters, above the wire floor of 16).
pub(crate) fn generate_nonce() -> Result<String, ConnectError> {
    let mut bytes = [0u8; 16];
    getrandom::getrandom(&mut bytes)
        .map_err(|e| ConnectError::Transport(format!("CSPRNG failure: {e}")))?;
    Ok(URL_SAFE_NO_PAD.encode(bytes))
}

/// The raw 32-byte Ed25519 seed of `identity` (the core's signing input).
///
/// Connect pins Ed25519 identity (protocol v1), so a differently keyed
/// identity is a configuration error.
fn ed25519_seed(identity: &Keypair) -> Result<[u8; 32], ConnectError> {
    let pair = identity
        .clone()
        .try_into_ed25519()
        .map_err(|_| ConnectError::Config("connect identity must be an Ed25519 keypair".into()))?;
    Ok(pair
        .secret()
        .as_ref()
        .try_into()
        .expect("libp2p Ed25519 secret key is 32 bytes"))
}

/// Sign a fresh hello for `identity` carrying `manifest` under `nonce`.
///
/// `peer_id` on the wire is derived from the Ed25519 public key (libp2p
/// PeerId base58btc multihash form), matching
/// `identity.public().to_peer_id().to_string()`.
pub(crate) fn sign_hello(
    identity: &Keypair,
    nonce: &str,
    manifest: &HostCapabilityManifest,
) -> Result<ConnectHello, ConnectError> {
    let seed = ed25519_seed(identity)?;
    core::sign_hello_ed25519(&seed, nonce, manifest).map_err(map_core_error)
}

/// Verify a received hello.
///
/// Checks, in order (all delegated to the core rule):
/// 1. `protocol_version` equals the protocol version (`HandshakeFailed`).
/// 2. `public_key` derives `expected_peer_id` — the noise-authenticated
///    remote peer (`HandshakeFailed`).
/// 3. The claimed `peer_id` equals `expected_peer_id` (`HandshakeFailed`).
/// 4. The signature verifies against `public_key` (`InvalidHelloSignature`).
///
/// Allowlist and nonce gates are applied by [`crate::gate`] around this check.
pub(crate) fn verify_hello(
    public_key: &PublicKey,
    expected_peer_id: &PeerId,
    hello: &ConnectHello,
) -> Result<(), ConnectError> {
    // The core rule operates on raw Ed25519 keys; connect pins Ed25519
    // identity, so any other remote key type fails the handshake closed.
    let pubkey =
        public_key
            .clone()
            .try_into_ed25519()
            .map_err(|_| ConnectError::HandshakeFailed {
                reason: "remote public key is not Ed25519".into(),
            })?;
    core::verify_hello_ed25519(&pubkey.to_bytes(), &expected_peer_id.to_string(), hello)
        .map_err(map_core_error)
}

#[cfg(test)]
mod tests {
    // Deterministic test vectors: the fixed nonce strings in these tests are
    // reproducible fixtures, not production CSPRNG output (production nonces
    // come from `generate_nonce`).
    use super::*;
    use crate::core::PROTOCOL_VERSION;
    use libp2p::identity::Keypair;

    fn manifest(host_id: &str) -> HostCapabilityManifest {
        HostCapabilityManifest {
            authority: None,
            capabilities: vec!["spoke-baseline".into()],
            extensions: Default::default(),
            host_id: host_id.parse().expect("host id parses"),
            namespaces: Vec::new(),
            roles: vec!["data-store".into()],
            schema_version: std::num::NonZeroU64::new(1).expect("non-zero"),
        }
    }

    #[test]
    fn nonce_meets_wire_floor_and_is_unique() {
        let a = generate_nonce().expect("nonce");
        let b = generate_nonce().expect("nonce");
        assert!(a.len() >= 16, "nonce below wire minLength: {}", a.len());
        assert_ne!(a, b);
    }

    #[test]
    fn short_caller_supplied_nonce_is_an_error_not_a_panic() {
        // The wire floor is 16 characters; a caller-supplied nonce below it
        // must surface as an error from the public signing API.
        let keypair = Keypair::generate_ed25519();
        let err = sign_hello(&keypair, "short", &manifest("host-a")).expect_err("short nonce");
        assert!(matches!(err, ConnectError::Config(_)));
    }

    #[test]
    fn sign_then_verify_round_trip() {
        let keypair = Keypair::generate_ed25519();
        let peer_id = keypair.public().to_peer_id();
        let hello =
            sign_hello(&keypair, "round-trip-nonce-123", &manifest("host-a")).expect("sign hello");

        assert_eq!(hello.protocol_version.get(), PROTOCOL_VERSION);
        assert_eq!(hello.peer_id.as_str(), peer_id.to_string());
        verify_hello(&keypair.public(), &peer_id, &hello).expect("verify hello");
    }

    #[test]
    fn tampered_host_fails_verification() {
        let keypair = Keypair::generate_ed25519();
        let peer_id = keypair.public().to_peer_id();
        let mut hello =
            sign_hello(&keypair, "tamper-nonce-12345", &manifest("host-a")).expect("sign hello");
        hello.host.roles.push("checker".into());

        let err = verify_hello(&keypair.public(), &peer_id, &hello).expect_err("tampered host");
        assert!(matches!(err, ConnectError::InvalidHelloSignature));
    }

    #[test]
    fn verify_key_must_derive_the_expected_peer_id() {
        // A key that does not derive the noise-authenticated peer id is
        // rejected at the binding check, before any signature work: the hello
        // signature must attest the authenticated peer's own identity key.
        let signer = Keypair::generate_ed25519();
        let other = Keypair::generate_ed25519();
        let hello =
            sign_hello(&signer, "wrong-key-nonce-12", &manifest("host-a")).expect("sign hello");

        let err = verify_hello(&other.public(), &signer.public().to_peer_id(), &hello)
            .expect_err("unbound verify key");
        assert!(
            matches!(err, ConnectError::HandshakeFailed { .. }),
            "unexpected error: {err:?}"
        );
    }

    #[test]
    fn claimed_peer_id_mismatch_fails_before_signature_check() {
        let keypair = Keypair::generate_ed25519();
        let other = Keypair::generate_ed25519();
        let hello =
            sign_hello(&keypair, "mismatch-nonce-123", &manifest("host-a")).expect("sign hello");

        // Verify with the signer's key but expect the *other* peer id: the
        // claimed peer id (from the hello) then disagrees with the expected one.
        let err = verify_hello(&keypair.public(), &other.public().to_peer_id(), &hello)
            .expect_err("peer id mismatch");
        assert!(matches!(err, ConnectError::HandshakeFailed { .. }));
    }

    #[test]
    fn unsupported_protocol_version_rejected() {
        let keypair = Keypair::generate_ed25519();
        let peer_id = keypair.public().to_peer_id();
        let mut hello =
            sign_hello(&keypair, "version-nonce-1234", &manifest("host-a")).expect("sign hello");
        hello.protocol_version = std::num::NonZeroU64::new(2).expect("non-zero");

        let err =
            verify_hello(&keypair.public(), &peer_id, &hello).expect_err("unsupported version");
        assert!(matches!(err, ConnectError::HandshakeFailed { .. }));
    }

    #[test]
    fn malformed_signature_rejected() {
        let keypair = Keypair::generate_ed25519();
        let peer_id = keypair.public().to_peer_id();
        let mut hello =
            sign_hello(&keypair, "bad-sig-nonce-123", &manifest("host-a")).expect("sign hello");
        hello.signature = "%%%not-base64url%%%"
            .parse()
            .expect("string parses as opaque");

        let err =
            verify_hello(&keypair.public(), &peer_id, &hello).expect_err("malformed signature");
        assert!(matches!(err, ConnectError::InvalidHelloSignature));
    }

    #[test]
    fn signed_fields_are_jcs_deterministic() {
        // Same fields, two keypairs → identical canonical bytes must sign and
        // verify independently (JCS determinism across signers).
        let a = Keypair::generate_ed25519();
        let b = Keypair::generate_ed25519();
        let manifest_a = manifest("host-x");
        let hello_a = sign_hello(&a, "jcs-nonce-1234567", &manifest_a).expect("sign a");
        let hello_b = sign_hello(&b, "jcs-nonce-1234567", &manifest_a).expect("sign b");

        // HostCapabilityManifest has no PartialEq derive; compare wire shape.
        assert_eq!(
            serde_json::to_value(&hello_a.host).expect("serialize a"),
            serde_json::to_value(&hello_b.host).expect("serialize b"),
        );
        assert_eq!(hello_a.nonce, hello_b.nonce);
        assert_ne!(hello_a.peer_id, hello_b.peer_id);

        verify_hello(&a.public(), &a.public().to_peer_id(), &hello_a).expect("a verifies");
        verify_hello(&b.public(), &b.public().to_peer_id(), &hello_b).expect("b verifies");
    }
}

File: packages/spoke-connect-ts/src/core/hello.ts
/**
 * Hello signing and verification over **raw** Ed25519 key bytes
 * (`spoke-connect-hello-jcs-v1`).
 *
 * Ported from `crates/spoke-connect/src/core/hello_crypto.rs`; normative
 * rules `.mstar/specs/spoke-connect.md` §Signature canonicalization /
 * §Identity binding:
 * - Signed object = exactly `{protocol_version, peer_id, nonce, host}` for
 *   the **initiator** hello (4 fields), and exactly
 *   `{protocol_version, peer_id, nonce, host, peer_nonce}` for the
 *   **responder** hello (5 fields, `peer_nonce` = the initiator's nonce —
 *   dial binding). Role is carried by the hello itself: `peer_nonce` is
 *   present in responder hellos, absent in initiator hellos.
 * - Canonicalize with RFC 8785 JCS → UTF-8 bytes (`src/jcs.ts`).
 * - Sign with the Ed25519 private key whose public key derives `peer_id`;
 *   the raw 64-byte signature is encoded base64url without padding.
 * - Top-level hello `extensions` and the `signature` field are **not** part
 *   of the signed object.
 *
 * Wire-shape gates enforced here (Rust enforces them via generated types):
 * `nonce` must be ≥ 16 chars and `protocol_version` must equal 1.
 */

import type { ConnectHello, HostCapabilityManifest } from "@42ch/spoke-schemas";

import {
  base64UrlDecode,
  base64UrlEncode,
  getPublicKeyEd25519,
  signEd25519,
  verifyEd25519,
} from "../crypto.js";
import { derivePeerIdFromEd25519Pubkey } from "../identity.js";
import { canonicalHelloBytes } from "../jcs.js";
import { CoreError } from "./error.js";
import { PROTOCOL_VERSION } from "./version.js";

/** JCS-canonicalize the signed hello object, mapping failures to `jcs`. */
function canonicalBytes(
  peerId: string,
  nonce: string,
  host: HostCapabilityManifest,
  peerNonce?: string,
): Uint8Array {
  try {
    return canonicalHelloBytes(peerId, nonce, host, peerNonce);
  } catch (error) {
    throw new CoreError(
      "jcs",
      error instanceof Error ? error.message : String(error),
    );
  }
}

/**
 * Sign a hello with a raw Ed25519 secret key (32-byte seed), producing the
 * full `ConnectHello` wire envelope.
 *
 * `peer_id` is derived from the public key of `secret`; the hello
 * `protocol_version` is the core protocol version. The `nonce` must meet the
 * wire floor (minLength 16) or `invalid_nonce` is thrown.
 *
 * Role: `peerNonce` is `undefined` for the **initiator** hello (4-field
 * signed object — the byte-identical pre-dial-binding path; the wire
 * envelope omits `peer_nonce` entirely) and the initiator's nonce for the
 * **responder** hello (5-field signed object, dial binding).
 */
export async function signHelloEd25519(
  secret: Uint8Array,
  nonce: string,
  manifest: HostCapabilityManifest,
  peerNonce?: string,
): Promise<ConnectHello> {
  if (secret.length !== 32) {
    throw new CoreError("crypto", "secret must be 32 bytes");
  }
  if (nonce.length < 16) {
    throw new CoreError("invalid_nonce", `nonce must be at least 16 characters (got ${nonce.length})`);
  }
  if (peerNonce !== undefined && peerNonce.length < 16) {
    throw new CoreError("invalid_nonce", `peer_nonce must be at least 16 characters (got ${peerNonce.length})`);
  }
  const publicKey = getPublicKeyEd25519(secret);
  const peerId = derivePeerIdFromEd25519Pubkey(publicKey);
  const bytes = canonicalBytes(peerId, nonce, manifest, peerNonce);
  const signature = base64UrlEncode(await signEd25519(secret, bytes));

  return {
    protocol_version: PROTOCOL_VERSION,
    peer_id: peerId,
    nonce,
    ...(peerNonce !== undefined ? { peer_nonce: peerNonce } : {}),
    host: manifest,
    signature,
    extensions: {},
  };
}

/**
 * Verify a received hello against a raw Ed25519 public key (32 bytes).
 *
 * Checks, in order:
 * 1. `protocol_version` equals the core protocol version
 *    (`protocol_version_mismatch` — a mixed-version peer or downgrade
 *    attempt; the dedicated kind, not `handshake_failed`).
 * 2. `nonce` meets the wire floor (minLength 16).
 * 3. The verify key derives `expectedPeerId` — the authenticated remote
 *    peer. A key that derives a different peer id cannot attest that peer's
 *    identity.
 * 4. The claimed `hello.peer_id` equals `expectedPeerId`.
 * 5. The signature verifies over the JCS-canonicalized signed object.
 *
 * Role-aware (dial binding): a hello carrying `peer_nonce` (responder) is
 * verified over the 5-field signed object, and when `expectedPeerNonce` is
 * supplied the initiator asserts the responder's `peer_nonce` equals its own
 * nonce — a replayed responder hello (signed over a different or absent
 * initiator nonce) fails here, which is the cross-restart replay defense.
 * Fail-closed: when `expectedPeerNonce` is supplied (the dial expects a
 * responder), a hello WITHOUT `peer_nonce` is rejected — the mixed-version
 * downgrade (an old responder omits the field) must not bypass the binding.
 * A hello without `peer_nonce` is verified over the 4-field signed object
 * only when no `expectedPeerNonce` is asserted.
 *
 * Allowlist and nonce gates are separate core checks (`allowlist` and
 * `nonce` modules).
 */
export async function verifyHelloEd25519(
  publicKey: Uint8Array,
  expectedPeerId: string,
  hello: ConnectHello,
  expectedPeerNonce?: string,
): Promise<void> {
  if (hello.protocol_version !== PROTOCOL_VERSION) {
    throw new CoreError(
      "protocol_version_mismatch",
      `unsupported protocol_version ${hello.protocol_version} (expected ${PROTOCOL_VERSION})`,
    );
  }
  if (hello.nonce.length < 16) {
    throw new CoreError("invalid_nonce", `hello nonce must be at least 16 characters (got ${hello.nonce.length})`);
  }
  if (hello.peer_nonce !== undefined && hello.peer_nonce.length < 16) {
    throw new CoreError("invalid_nonce", `hello peer_nonce must be at least 16 characters (got ${hello.peer_nonce.length})`);
  }
  if (publicKey.length !== 32) {
    throw new CoreError("crypto", "public key must be 32 bytes");
  }
  const derivedPeerId = derivePeerIdFromEd25519Pubkey(publicKey);
  if (derivedPeerId !== expectedPeerId) {
    throw new CoreError(
      "handshake_failed",
      `public key derives peer id ${derivedPeerId} instead of the authenticated peer ${expectedPeerId}`,
    );
  }
  if (hello.peer_id !== expectedPeerId) {
    throw new CoreError(
      "handshake_failed",
      `hello peer_id ${hello.peer_id} does not match authenticated peer ${expectedPeerId}`,
    );
  }

  // The claimed peer id was just checked to equal `expectedPeerId`, so
  // canonicalizing over `expectedPeerId` reproduces the signer's bytes.
  // Role-aware signed object: 5 fields (incl. `peer_nonce`) for a responder
  // hello, 4 fields for an initiator hello. When the initiator supplies
  // `expectedPeerNonce`, the responder's `peer_nonce` must be PRESENT AND
  // equal to it — this is the dial binding that defeats cross-restart
  // replay. Fail-closed (spec §Dial binding): a dial that expects a
  // responder MUST NOT accept an initiator-shaped hello — a mixed-version
  // old responder omits `peer_nonce`, and silently falling back to the
  // 4-field verify would skip the binding assert entirely.
  let bytes: Uint8Array;
  if (hello.peer_nonce !== undefined) {
    if (
      expectedPeerNonce !== undefined &&
      hello.peer_nonce !== expectedPeerNonce
    ) {
      throw new CoreError(
        "handshake_failed",
        `responder hello peer_nonce does not match the initiator nonce (dial binding)`,
      );
    }
    bytes = canonicalBytes(
      expectedPeerId,
      hello.nonce,
      hello.host,
      hello.peer_nonce,
    );
  } else {
    if (expectedPeerNonce !== undefined) {
      throw new CoreError(
        "handshake_failed",
        "responder hello is missing peer_nonce (dial binding)",
      );
    }
    bytes = canonicalBytes(expectedPeerId, hello.nonce, hello.host);
  }

  let signature: Uint8Array;
  try {
    signature = base64UrlDecode(hello.signature);
  } catch {
    throw new CoreError("invalid_hello_signature");
  }
  if (signature.length !== 64) {
    throw new CoreError("invalid_hello_signature");
  }
  const ok = await verifyEd25519(publicKey, bytes, signature);
  if (!ok) {
    throw new CoreError("invalid_hello_signature");
  }
}
